diff -Naur os/pkg/init/fsmount/fsmount.go os.patched/pkg/init/fsmount/fsmount.go
--- os/pkg/init/fsmount/fsmount.go	2023-11-17 14:32:19.409687529 +0900
+++ os.patched/pkg/init/fsmount/fsmount.go	2023-11-17 23:11:53.542464794 +0900
@@ -1,9 +1,13 @@
 package fsmount
 
 import (
+	"bufio"
+	"bytes"
 	"fmt"
 	"strings"
 	"syscall"
+	"os"
+	"os/exec"
 
 	"github.com/burmilla/os/config"
 	"github.com/burmilla/os/config/cmdline"
@@ -22,7 +26,7 @@
 )
 
 func MountOem(cfg *config.CloudConfig) (*config.CloudConfig, error) {
-	if err := mountConfigured("oem", cfg.Rancher.State.OemDev, cfg.Rancher.State.OemFsType, config.OemDir); err != nil {
+	if err := mountConfigured("oem", cfg.Rancher.State.OemDev, cfg.Rancher.State.OemFsType, config.OemDir, cfg.Rancher.State.OemMountOpt); err != nil {
 		log.Debugf("Not mounting OEM: %v", err)
 	} else {
 		log.Infof("Mounted OEM: %s", cfg.Rancher.State.OemDev)
@@ -45,7 +49,7 @@
 		}
 	}
 
-	if err := mountConfigured("boot", cfg.Rancher.State.BootDev, cfg.Rancher.State.BootFsType, config.BootDir); err != nil {
+	if err := mountConfigured("boot", cfg.Rancher.State.BootDev, cfg.Rancher.State.BootFsType, config.BootDir, cfg.Rancher.State.BootMountOpt); err != nil {
 		log.Debugf("Not mounting BOOT: %v", err)
 	} else {
 		log.Infof("Mounted BOOT: %s", cfg.Rancher.State.BootDev)
@@ -54,7 +58,37 @@
 	return cfg, nil
 }
 
-func mountConfigured(display, dev, fsType, target string) error {
+func blkidMatchStr(matchStr string) (deviceName, deviceType string, err error) {
+	cmd := exec.Command("/sbin/blkid")
+	cmd.Stderr = os.Stderr
+	out, err := cmd.Output()
+	if err != nil {
+		fmt.Println("exec /sbin/blkid, failed.")
+		fmt.Println(err)
+		return
+	}
+	r := bytes.NewReader(out)
+	s := bufio.NewScanner(r)
+	for s.Scan() {
+		line := s.Text()
+		if !strings.Contains(line, matchStr) {
+			continue
+		}
+		d := strings.Split(line, ":")
+		deviceName = d[0]
+
+		s1 := strings.Split(line, `TYPE="`)
+		s2 := strings.Split(s1[1], `"`)
+		deviceType = s2[0]
+		return
+	}
+	fmt.Println("extract output from /sbin/blkid, failed.")
+	fmt.Println(string(out))
+	return
+}
+
+
+func mountConfigured(display, dev, fsType, target, option string) error {
 	var err error
 
 	if dev == "" {
@@ -63,7 +97,13 @@
 
 	dev = util.ResolveDevice(dev)
 	if dev == "" {
-		return fmt.Errorf("Could not resolve device %q", dev)
+		d, _, err := blkidMatchStr(dev)
+		if err != nil {
+			return fmt.Errorf("Could not resolve device %q", dev)
+		}
+		if d == "" {
+			return fmt.Errorf("Could not resolve device %q", dev)
+		}
 	}
 	if fsType == "auto" {
 		fsType, err = util.GetFsType(dev)
@@ -74,12 +114,12 @@
 	}
 
 	log.Debugf("FsType has been set to %s", fsType)
-	log.Infof("Mounting %s device %s to %s", display, dev, target)
-	return util.Mount(dev, target, fsType, "")
+	log.Infof("Mounting %s device %s to %s (option = '%s')", display, dev, target, option)
+	return util.Mount(dev, target, fsType, option)
 }
 
 func mountState(cfg *config.CloudConfig) error {
-	return mountConfigured("state", cfg.Rancher.State.Dev, cfg.Rancher.State.FsType, config.StateDir)
+	return mountConfigured("state", cfg.Rancher.State.Dev, cfg.Rancher.State.FsType, config.StateDir, cfg.Rancher.State.MountOpt)
 }
 
 func tryMountState(cfg *config.CloudConfig) error {
diff -Naur os/config/types.go os.patched/config/types.go
--- os/config/types.go	2023-11-17 14:32:19.394687792 +0900
+++ os.patched/config/types.go	2023-11-17 23:01:31.863279855 +0900
@@ -210,6 +210,7 @@
 type StateConfig struct {
 	Directory  string   `yaml:"directory,omitempty"`
 	FsType     string   `yaml:"fstype,omitempty"`
+	MountOpt     string   `yaml:"mountopt,omitempty"`
 	Dev        string   `yaml:"dev,omitempty"`
 	Wait       bool     `yaml:"wait,omitempty"`
 	Required   bool     `yaml:"required,omitempty"`
@@ -220,8 +221,10 @@
 	Rngd       bool     `yaml:"rngd,omitempty"`
 	Script     string   `yaml:"script,omitempty"`
 	OemFsType  string   `yaml:"oem_fstype,omitempty"`
+	OemMountOpt  string   `yaml:"oem_mountopt,omitempty"`
 	OemDev     string   `yaml:"oem_dev,omitempty"`
 	BootFsType string   `yaml:"boot_fstype,omitempty"`
+	BootMountOpt string   `yaml:"boot_mountopt,omitempty"`
 	BootDev    string   `yaml:"boot_dev,omitempty"`
 }
 
diff -Naur os/scripts/version os.patched/scripts/version
--- os/scripts/version	2023-11-17 14:32:19.427687212 +0900
+++ os.patched/scripts/version	2023-11-17 23:59:30.648734315 +0900
@@ -1,18 +1,26 @@
 #!/bin/bash
 
-if [ -n "$(git status --porcelain --untracked-files=no)" ]; then
-    DIRTY="-dirty"
-fi
+#if [ -n "$(git status --porcelain --untracked-files=no)" ]; then
+#    DIRTY="-dirty"
+#fi
 
 COMMIT=$(git rev-parse --short HEAD)
 GIT_TAG=$(git tag -l --contains HEAD | head -n 1)
 
-if [ -z "$VERSION" ]; then
-    if [[ -z "$DIRTY" && -n "$GIT_TAG" ]]; then
-        VERSION=$GIT_TAG
-    else
-        VERSION="${COMMIT}${DIRTY}"
-    fi
+#if [ -z "$VERSION" ]; then
+#    if [[ -z "$DIRTY" && -n "$GIT_TAG" ]]; then
+#        VERSION=$GIT_TAG-btrfs-0.9
+#    else
+#        VERSION="${COMMIT}${DIRTY}"
+#    fi
+#fi
+
+if [ ! -e ../.git/packed-refs ]; then
+	VERSION=`cat /tmp/version.patched`
+else
+	TTAG=`tail -1 ../.git/packed-refs`
+	VERSION=${TTAG##*/}-btrfs-__PVER__
+	echo -n $VERSION > /tmp/version.patched
 fi
 
 INITRD=initrd-${VERSION}
